src/sentry/sentry_metrics/indexer/cache.py

```diff
@@ -82,3 +95,77 @@ def delete(self, key: str, cache_namespace: str) -> None:
     def delete_many(self, keys: Sequence[str], cache_namespace: str) -> None:
         cache_keys = [self.make_cache_key(key, cache_namespace) for key in keys]
         self.cache.delete_many(cache_keys, version=self.version)
+
+
+class CachingIndexer(StringIndexer):
+    def __init__(self, cache: StringIndexerCache, indexer: StringIndexer) -> None:
+        self.cache = cache
+        self.indexer = indexer
+
+    def bulk_record(
+        self, use_case_id: UseCaseKey, org_strings: Mapping[int, Set[str]]
+    ) -> KeyResults:
+        cache_keys = KeyCollection(org_strings)
+        metrics.gauge("sentry_metrics.indexer.lookups_per_batch", value=cache_keys.size)
+        cache_key_strs = cache_keys.as_strings()
+        cache_results = self.cache.get_many(cache_key_strs, use_case_id.value)
+
+        hits = [k for k, v in cache_results.items() if v is not None]
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "true", "caller": "get_many_ids"},
+            amount=len(hits),
+        )
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "false", "caller": "get_many_ids"},
+            amount=len(cache_results) - len(hits),
+        )
+
+        # used to compare to pre org_id indexer cache fetch metric
+        metrics.incr(
+            _INDEXER_CACHE_FETCH_METRIC,
+            amount=cache_keys.size,
+        )
+
+        cache_key_results = KeyResults()
+        cache_key_results.add_key_results(
+            [KeyResult.from_string(k, v) for k, v in cache_results.items() if v is not None],
+            FetchType.CACHE_HIT,
+        )
+
+        db_record_keys = cache_key_results.get_unmapped_keys(cache_keys)
+
+        if db_record_keys.size == 0:
+            return cache_key_results
+
+        db_record_key_results = self.indexer.bulk_record(use_case_id, db_record_keys.mapping)
+        self.cache.set_many(
```

Comment:
is this going to set the rate-limited values with `id=None` in the cache? Do we want that?

Reply:
good catch. I think i unintentionally changed this and now i'm passing a keyresult that contains rate limited strings as well. but as you said seems like it all works out fine in the end